### PR TITLE
add all_mailinglists and fetch_mailinglist functions

### PR DESCRIPTION
**Pull request by dsen6644**
Mailing list functionality for research core (does not support xm directories)

`all_mailinglists` will return a data frame of all mailing lists with their respective library ids, name, and folder structure.

`fetch_mailinglist` will return the mailing list in data frame format.

### CODE REVIEW

**Comment by juliasilge** [comment]
This looks fantastic @dsen6644! Would you be interested in adding yourself as a contributor (`ctb`) in DESCRIPTION?

I have a bigger question to ask you. Have you ever had experience in testing API packages? Or are you interested in learning a bit about how to do it? I want to revamp the whole testing suite, as I mention in #120, but we wouldn't have to do it all at once. If you wrote a couple of unit tests for these new functions, we could start shifting over to the new framework gradually. I would like to use [vcr](https://github.com/ropensci/vcr).

**Comment by dsen6644** [comment]
I do not have any experience with api tests, but very interested in learning. I'll look into the `vcr` package, but may wait for you to commit your first test so I can follow your style/format.

**Comment by juliasilge** [comment]
OK! I am about to submit to CRAN (just waiting on win-builder to come back) and after that I'll work on getting some tests started using vcr (starting with `all_surveys()`, probably). I'd like to get a couple of tests on this branch for the mailing list functions before I merge it in.

If you want to just go for it, by all means do! 💪

**Comment by juliasilge** [comment]
Hello there @dsen6644! I am super happy with this and would love to merge it in; it just needs a couple of tests for the new functions. I haven't gotten a chance to switch over old tests to vcr yet, but [this resource from rOpenSci](https://books.ropensci.org/http-testing/) is pretty extensive if you want to get started.

Or if you are happy to wait until we can get older tests moved over, that is fine too. 🙌

File: R/fetch_mailinglist.R [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  assert_api_key()
+
+  fetch_url <- create_mailinglist_url(Sys.getenv("QUALTRICS_BASE_URL"), mailinglistID)
+
+  elements <- list()
+
+  while(!is.null(fetch_url)){
+
+    res <- qualtrics_api_request("GET", url = fetch_url)
+    elements <- append(elements, res$result$elements)
```
Comment (jmobrien) [inline R/fetch_mailinglist.R]: `res$result$elements` is a list object here, is it not?  If so, I'm not sure you'd end up with the intended `elements` final list if you have >1 pass through this loop.  See, for example:

```r
test <- list(a = 1:10, b = -1:-10)
append(test, test)
```

**Comment by jmobrien** [comment]
Hi, great work @dsen6644!  Glad to see someone adding this feature in formally; hacking in something like this was how I actually got started using the rest of qualtRics as much as I do now.

Something I fought with back then, though: in both of these new functions, you might want to check that the while loops are properly configured to stack the data, in case you have >1 pass of data from the API   Think it needs sub-element-wise stacking.  I made a comment to the commit to that effect.

(I do have a branch "Mailinglist" on my page where you can see what I actually did with the above. But that's just FYI if it's useful; it's just old stuff copied up, not clearly an improvement otherwise on this nice work.  Love that you're taking the lead on this useful feature, so I don't want to get in your way.)

File: R/fetch_mailinglist.R (hunk 1 above)
Comment (jmobrien) [reply R/fetch_mailinglist.R:27]: Ah I understand what you're doing better now!  You're completely right, and my example was way off the mark even before your clarification; I'd forgotten that the data returned from this endpoint is formatted as elements at the case level.  My mistake, and sorry for wasting your time!

(Looks like I even read my own code wrong, too!  Pretty obvious looking back at it I was just starting with purrr and didn't really grasp it well at all.  I'm glad you are coming at this with a better sense.)

Comment (jmobrien) [reply R/fetch_mailinglist.R:27]: Related to the subsequent binding, though, in trying to figure out my own code from earlier I stumbled on a possible way to preserve `responseHistory` and `emailHistory` elements as list-columns.  With pipes for ease of reading:
```r
elements %>% 
  purrr::map(
    ~purrr::modify_if(.x, purrr::is_empty, ~NA) %>% 
      purrr::modify_if(is.list, list)
  ) %>% 
  bind_rows %>% 
  tidyr::unnest_wider(col = embeddedData)
```
Not sure how desirable including those would be; it preserves data but the resulting dataframe isn't as easily writeable. 

(Also, question: I'm noticing that pretty much everything is coming back from the endpoint as character-mode already, even my numeric ID codes.  Are you seeing the same?  I didn't upload this specific list I'm looking at so I'm not sure what the default behavior is.)

Comment (dsen6644) [reply R/fetch_mailinglist.R:27]: I thought about preserving `responseHistory` and `emailHistory` as list columns, but to your point, the resulting object isn't as clean. That and I felt that response and email history should probably be handled by functions geared towards fetching distribution history (which I'm currently working on).

Overall I think the function should mimic the output you'd see if you download the contact list from Qualtrics directly.

Comment (dsen6644) [reply R/fetch_mailinglist.R:27]: And yes the objects I'm seeing returned are of type character.

**Comment by jmobrien** [comment]
Makes sense, especially if this matches the web interface download, which
I'm not familiar with.  Thanks again for this great work!

Joseph O'Brien, Ph.D.
Adolescent Development Research Group
Department of Psychology
University of Texas at Austin
joseph.obrien@utexas.edu

**Comment by juliasilge** [comment]
Thank you so much for starting the vcr testing @dsen6644!!! I believe you also need to add the `fixtures` directory to the repo. It might help to check out how this is implemented in [ropenaq](https://github.com/ropensci/ropenaq/tree/master/tests).

**Comment by juliasilge** [comment]
Also @dsen6644 would you like to add yourself to `DESCRIPTION` as a contributor?

**Comment by dsen6644** [comment]
Are the URLs being generated by our .util functions `vcr` friendly? Or have I missed a step? Same failure for both `all_surveys` and `all_mailinglists`.

```
vcr::use_cassette("all_mailinglists", {
  x <- all_mailinglists()
})
#> CrulAdapter enabled!
#> HttrAdapter enabled!
#> net connect allowed
#> Error in curl::curl_fetch_memory(url, handle = handle): Bad URL, colon is first character
#> ejecting cassette: all_mailinglists
#> CrulAdapter disabled!
#> HttrAdapter disabled!

vcr::use_cassette("all_surveys", {
  x <- all_surveys()
})
#> CrulAdapter enabled!
#> HttrAdapter enabled!
#> Error in curl::curl_fetch_memory(url, handle = handle): Bad URL, colon is first character
#> ejecting cassette: all_surveys
#> CrulAdapter disabled!
#> HttrAdapter disabled!
```

vcr configuration

```
vcr::vcr_config_defaults()
#> $dir
#> [1] "."
#> 
#> $record
#> [1] "once"
#> 
#> $match_requests_on
#> [1] "method" "uri"   
#> 
#> $allow_unused_http_interactions
#> [1] TRUE
#> 
#> $serialize_with
#> [1] "yaml"
#> 
#> $persist_with
#> [1] "FileSystem"
#> 
#> $ignore_hosts
#> NULL
#> 
#> $ignore_localhost
#> [1] FALSE
#> 
#> $ignore_request
#> NULL
#> 
#> $uri_parser
#> [1] "crul::url_parse"
#> 
#> $preserve_exact_body_bytes
#> [1] FALSE
#> 
#> $turned_off
#> [1] FALSE
#> 
#> $re_record_interval
#> NULL
#> 
#> $clean_outdated_http_interactions
#> NULL
#> 
#> $allow_http_connections_when_no_cassette
#> [1] FALSE
#> 
#> $cassettes
#> list()
#> 
#> $linked_context
#> NULL
#> 
#> $log
#> [1] FALSE
#> 
#> $log_opts
#> $log_opts$file
#> [1] "vcr.log"
#> 
#> $log_opts$log_prefix
#> [1] "Cassette"
#> 
#> 
#> $filter_sensitive_data
#> NULL
#> 
#> $write_disk_path
#> NULL
```

**Comment by dsen6644** [comment]
adding `https://` to the root url seems to fix this, will need to update url functions to work with vcr package.

**Comment by dsen6644** [comment]
When running `all_mailinglists` within `vcr::use_cassette` it records (as expected) the request and returned elements, which includes my mailing list names in the body.

I'm already using the `filter_sensitive_data` argument in the vcr_configure to protect my base_url and api key, but I don't see anything in the documentation around filtering out elements within the body? i.e I want to record the structure of the data frame for testing, but not the names of my personal mailing lists.

See the excerpt of the .yml file below, with my library id value being substituted with `LIBRARY-ID-VALUE` manually.

```
http_interactions:
- request:
    method: get
    uri: https://<<<my_base_url>>>/API/v3/mailinglists/
    body:
      encoding: ''
      string: ''
    headers:
      X-API-TOKEN: <<<my_api_key>>>
      Content-Type: application/json
      Accept: '*/*'
      accept-encoding: gzip, deflate
  response:
    status:
      status_code: 200
      category: Success
      reason: OK
      message: 'Success: (200) OK'
    headers:
      content-length: '4429'
      content-type: application/json
      request-time: '180'
      server: devplat-gateway
      content-encoding: gzip
      date: Sun, 26 Apr 2020 23:14:08 GMT
      connection: keep-alive
      vary: Accept-Encoding
      content-security-policy-report-only: frame-ancestors 'self' *.qualtrics.com;
        report-uri https://sjc1.qualtrics.com/csp-report
      strict-transport-security: max-age=31536000; includeSubDomains; preload
    body:
      encoding: UTF-8
      file: no
      string: '{"result":{"elements":[{"libraryId":"**<LIBRARY-ID-VALUE>**"
```

**Comment by juliasilge** [comment]
Ah yes, I think that the vcr library is set up to actually record the data from the API call; this is what the tests will use to check against. I haven't used mailing lists in Qualtrics much, but is it possible for you to [create a small list with fake emails](https://fauxid.com/tools/fake-email-list) in it?

**Comment by dsen6644** [comment]
That would work for `fetch_mailinglist` but not `all_mailinglists`, which would, unfortunately, return the names of all my mailing lists, some of which are sensitive. How would you handle a similar issue for `all_surveys`?

**Comment by juliasilge** [comment]
Good point, of course. Maybe I should do it on the free account Qualtrics gave me where I don't have any actual sensitive data.

Can you set this branch up as much as you can, _except_ for committing the actual files with API results? Try to get it passing R CMD check with tests locally for you. I will fetch this branch and try to finish it off. If you can make me a TODO list, that would be especially helpful!

**Comment by dsen6644** [comment]
This branch is now up to date with tests for `all_mailinglists` and `fetch_mailinglist`. The package passes all checks locally when my api credentials are used.

TODO
- Edit dummy api credentials in `test-fetch-mailinglist` and `test-all-mailinglists`
- Add valid mailinglist id in `fetch_mailinglist` function in `test-fetch-mailinglist`
- Run VCR functions independently of check to store .yml file

Note: Unlike some of the survey response function tests, I did not add any tests around the authentication of api credentials. I figured our util functions should have their own, separate, vcr tests.

Note2: I'm still trying to discern the difference between `devtools::check` and `devtools::test` as discussed [here](https://books.ropensci.org/http-testing/vcr-intro.html#check-vs-test).

File: R/fetch_mailinglist.R [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  x <- tibble::tibble(id = purrr::map_chr(elements, "id", .default = NA_character_),
+                      firstName = purrr::map_chr(elements, "firstName", .default = NA_character_),
+                      lastName = purrr::map_chr(elements, "lastName", .default = NA_character_),
+                      email = purrr::map_chr(elements, "email", .default = NA_character_),
+                      externalDataReference = purrr::map_chr(elements, "externalDataReference", .default = NA_character_),
+                      language = purrr::map_chr(elements, "language", .default = NA_character_),
+                      unsubscribed = purrr::map_lgl(elements, "unsubscribed", .default = NA))
+
+  x <- dplyr::bind_cols(x, purrr::map_df(elements, "embeddedData", .default = NA_character_))
```
Comment (juliasilge) [inline R/fetch_mailinglist.R:49]: This line does not work for the mailing lists I created in Qualtrics. I get this error:

```
Error: Argument 1 must have names
Backtrace:
 1. dplyr::bind_cols(...)
 5. purrr::map_df(elements, "embeddedData", .default = NA_character_)
 6. dplyr::bind_rows(res, .id = .id)
 7. dplyr:::bind_rows_(x, .id)
```

I'm about to push a change that works for me; can you pull and see if the new version works for you?

**Comment by juliasilge** [comment]
@dsen6644 did you successfully have your API key and base URL filtered / replaced in the fixtures that were being written to disk? They are not being replaced in the ones I am creating.

File: R/fetch_mailinglist.R (hunk 2 above)
Comment (dsen6644) [reply R/fetch_mailinglist.R]: Yes, the new commit works for all my mailing lists. `enframe` is a much safer alternative to `map_df`.

**Comment by dsen6644** [comment]
It works when I run `devtools::test()`the first time... anytime after that, I get the following-

```
An HTTP request has been made that vcr does not know how to handle:
GET <<<my_base_url>>><<<<my_base_url>>><<<<my_base_url>>><<<<my_base_url>>><<<<my_base_url>>><<<<my_base_url>>><<<<my_base_url>>><<<<my_base_url>>><<<<my_base_url>>><<<<my_base_url>>><<<<my_base_url>>>.....
```

**Comment by jrosen48** [comment]
Hi all, excited to see all of the development taking place around an already-great package. @dsen6644 I reached out to @juliasilge via Twitter about helping out (https://twitter.com/juliasilge/status/1257004429921759232) - and just wanted to follow-up to say I'd be happy to serve as another tester/to try to help out on any vcr testing-related additions. I will continue to follow along for now (and will see if there's a spot where I can help).

**Comment by juliasilge** [comment]
All changes HAVE PASSED!!! 🎉 

GIPHY

It only took THREE MONTHS!!!! 😅 But think how much we learned in the process!

**Comment by juliasilge** [comment]
Between this and #161 we went from ~50% to ~80% testing coverage today! 💥

**Pull request merged**